### PR TITLE
Make document symbols more resilient

### PR DESCRIPTION
**Pull request by lukaszsamson**
Do not crash on incomplete typespecs

### CODE REVIEW

File: apps/language_server/lib/language_server/providers/document_symbols.ex [hunk 1]
```diff
@@ -169,14 +169,18 @@ defmodule ElixirLS.LanguageServer.Providers.DocumentSymbols do
 
   # Types
   defp extract_symbol(_current_module, {:@, location, [{type_kind, _, type_expression}]})
-       when type_kind in [:type, :typep, :opaque, :callback, :macrocallback] do
+       when type_kind in [:type, :typep, :opaque, :callback, :macrocallback] and
```
Comment (scohen) [inline apps/language_server/lib/language_server/providers/document_symbols.ex:172]: this is getting pretty complicated, can we do the nill check in a separate function head before this one?

```elixir
defp extract_symbol(_current_module, {:@, location, [{type_kind, _, nil}]}) do
  ...
end
```

File: apps/language_server/test/providers/document_symbols_test.exs [hunk 2]
```diff
@@ -1216,6 +1218,16 @@ defmodule ElixirLS.LanguageServer.Providers.DocumentSymbolsTest do
                     children: [],
                     kind: 5,
                     name: "@type my_with_args_when(key, value)"
+                  },
+                  %Protocol.DocumentSymbol{
+                    children: [],
+                    kind: 5,
```
Comment (scohen) [inline apps/language_server/test/providers/document_symbols_test.exs:1224]: man, I really dislike these magic numbers...
one idea is to use the experimental protocol structs. Another is to define functions in `Protocol.DocumentSymbol` named for the kinds that fill them in for you. i.e.
```elixir
Protocol.DocumentSymbol.class(name: "@type abc")
```

**Review by scohen — APPROVED** [review]
approved with considerations

File: apps/language_server/lib/language_server/providers/document_symbols.ex (hunk 1 above)
Comment (lukaszsamson) [reply apps/language_server/lib/language_server/providers/document_symbols.ex:172]: Then we would have attribute handling in 2 clauses - one with `nil` and one lower matching `_`. Rejecting `nil` here make it fall back to the lower `_` clause handling all non typespec attributes

File: apps/language_server/test/providers/document_symbols_test.exs (hunk 2 above)
Comment (lukaszsamson) [reply apps/language_server/test/providers/document_symbols_test.exs:1224]: Extracted to a separate issue https://github.com/elixir-lsp/elixir-ls/issues/831

**Pull request merged**